File: Proyecto_edd/ArbolAVL.java
public class ArbolAVL {
    private Nodo raiz;

    // Método para insertar una tienda en el árbol
    public void insertar(Tienda tienda) {
        if (tienda == null) {
            System.out.println("No se puede insertar una tienda nula.");
            return;
        }
        raiz = insertar(raiz, tienda);
    }

    private Nodo insertar(Nodo nodo, Tienda tienda) {
        if (nodo == null) {
            return new Nodo(tienda);
        }

        if (tienda.getIdTienda() < nodo.tienda.getIdTienda()) {
            nodo.izquierda = insertar(nodo.izquierda, tienda);
        } else if (tienda.getIdTienda() > nodo.tienda.getIdTienda()) {
            nodo.derecha = insertar(nodo.derecha, tienda);
        } else {
            // No se permiten claves duplicadas
            return nodo;
        }

        // Actualizar altura del nodo
        nodo.altura = 1 + Math.max(altura(nodo.izquierda), altura(nodo.derecha));

        // Balancear el nodo
        return balancear(nodo, tienda.getIdTienda());
    }

    private int altura(Nodo nodo) {
        return nodo == null ? 0 : nodo.altura;
    }

    private int obtenerBalance(Nodo nodo) {
        return nodo == null ? 0 : altura(nodo.izquierda) - altura(nodo.derecha);
    }

    private Nodo balancear(Nodo nodo, int clave) {
        int balance = obtenerBalance(nodo);

        // Rotación simple a la derecha
        if (balance > 1 && clave < nodo.izquierda.tienda.getIdTienda()) {
            return rotarDerecha(nodo);
        }

        // Rotación simple a la izquierda
        if (balance < -1 && clave > nodo.derecha.tienda.getIdTienda()) {
            return rotarIzquierda(nodo);
        }

        // Rotación doble a la derecha
        if (balance > 1 && clave > nodo.izquierda.tienda.getIdTienda()) {
            nodo.izquierda = rotarIzquierda(nodo.izquierda);
            return rotarDerecha(nodo);
        }

        // Rotación doble a la izquierda
        if (balance < -1 && clave < nodo.derecha.tienda.getIdTienda()) {
            nodo.derecha = rotarDerecha(nodo.derecha);
            return rotarIzquierda(nodo);
        }

        return nodo;
    }

    private Nodo rotarDerecha(Nodo y) {
        Nodo x = y.izquierda;
        Nodo T2 = x.derecha;

        // Rotar
        x.derecha = y;
        y.izquierda = T2;

        // Actualizar alturas
        y.altura = Math.max(altura(y.izquierda), altura(y.derecha)) + 1;
        x.altura = Math.max(altura(x.izquierda), altura(x.derecha)) + 1;

        return x;
    }

    private Nodo rotarIzquierda(Nodo x) {
        Nodo y = x.derecha;
        Nodo T2 = y.izquierda;

        // Rotar
        y.izquierda = x;
        x.derecha = T2;

        // Actualizar alturas
        x.altura = Math.max(altura(x.izquierda), altura(x.derecha)) + 1;
        y.altura = Math.max(altura(y.izquierda), altura(y.derecha)) + 1;

        return y;
    }

    // Método para buscar la tienda más cercana
    public Tienda buscarCercano(String ubicacionUsuario) {
        // Simulación de búsqueda. Podrías implementar una métrica de distancia real.
        return raiz != null ? raiz.tienda : null;
    }
}

File: proyecto_version_light/scytec/src/ArbolAVL.py
class NodoAVL:
    def __init__(self, clave, dato):
        self.clave = clave
        self.dato = dato
        self.izquierda = None
        self.derecha = None
        self.altura = 1


class ArbolAVL:
    def __init__(self):
        self.raiz = None

    def insertar(self, dato):
        self.raiz = self._insertar(self.raiz, dato)

    def _insertar(self, nodo, dato):
        if not nodo:
            return NodoAVL(dato[0], dato[1])
        if dato[0] < nodo.clave:
            nodo.izquierda = self._insertar(nodo.izquierda, dato)
        else:
            nodo.derecha = self._insertar(nodo.derecha, dato)
        nodo.altura = 1 + max(self._altura(nodo.izquierda), self._altura(nodo.derecha))
        return self._balancear(nodo)

    def en_orden(self):
        resultado = []
        self._en_orden(self.raiz, resultado)
        return resultado

    def _en_orden(self, nodo, resultado):
        if nodo:
            self._en_orden(nodo.izquierda, resultado)
            resultado.append((nodo.clave, nodo.dato))
            self._en_orden(nodo.derecha, resultado)

    def _altura(self, nodo):
        return nodo.altura if nodo else 0

    def _balancear(self, nodo):
        balance = self._altura(nodo.izquierda) - self._altura(nodo.derecha)
        if balance > 1:
            if self._altura(nodo.izquierda.izquierda) >= self._altura(nodo.izquierda.derecha):
                return self._rotar_derecha(nodo)
            else:
                nodo.izquierda = self._rotar_izquierda(nodo.izquierda)
                return self._rotar_derecha(nodo)
        if balance < -1:
            if self._altura(nodo.derecha.derecha) >= self._altura(nodo.derecha.izquierda):
                return self._rotar_izquierda(nodo)
            else:
                nodo.derecha = self._rotar_derecha(nodo.derecha)
                return self._rotar_izquierda(nodo)
        return nodo

    def _rotar_derecha(self, nodo):
        nueva_raiz = nodo.izquierda
        nodo.izquierda = nueva_raiz.derecha
        nueva_raiz.derecha = nodo
        nodo.altura = 1 + max(self._altura(nodo.izquierda), self._altura(nodo.derecha))
        nueva_raiz.altura = 1 + max(self._altura(nueva_raiz.izquierda), self._altura(nueva_raiz.derecha))
        return nueva_raiz

    def _rotar_izquierda(self, nodo):
        nueva_raiz = nodo.derecha
        nodo.derecha = nueva_raiz.izquierda
        nueva_raiz.izquierda = nodo
        nodo.altura = 1 + max(self._altura(nodo.izquierda), self._altura(nodo.derecha))
        nueva_raiz.altura = 1 + max(self._altura(nueva_raiz.izquierda), self._altura(nueva_raiz.derecha))
        return nueva_raiz
